crates/solver/src/liquidity/zeroex.rs

```diff
@@ -9,39 +9,287 @@ use {
 };
 
+const DEFAULT_ZEROEX_WEBSOCKET_API: &str = "wss://api.0x.org/orderbook/v1";
+
+type Socket = WebSocketStream<MaybeTlsStream<TcpStream>>;
+
+#[derive(Clone)]
+pub struct ZeroExWebsocketApi {
+    url: Url,
+}
+
+impl ZeroExWebsocketApi {
+    pub fn new(url: impl IntoUrl) -> Self {
+        Self {
+            url: url
+                .into_url()
+                .unwrap_or(Url::parse(DEFAULT_ZEROEX_WEBSOCKET_API).unwrap()),
+        }
+    }
+}
+
+impl ZeroExWebsocketApi {
+    /// Creates a websocket connection to 0x and subscribes to orderbook
+    /// updates. Once this function is done, the websocket is open and ready
+    /// to receive messages.
+    async fn open_socket(&self) -> Result<Socket> {
+        let mut socket = tokio_tungstenite::connect_async(self.url.clone()).await?.0;
+
+        // Construct the subscription message to fetch all orders
+        let subscription_msg = serde_json::json!({
+            "type": "subscribe",
+            "channel": "orders",
+            "requestId": "cowswap",
+        });
+
+        socket
+            .send(Message::Text(subscription_msg.to_string()))
+            .await?;
+
+        Ok(socket)
+    }
+}
+
+#[derive(Debug)]
+enum UpdateError {
+    // The websocket is closed on server side, with optional reason
+    SocketClosed(Option<anyhow::Error>),
+    // Received a message type that we don't handle
+    UnsupportedMessage,
+    // Received proper `text` type of message but it can't be deserialized (mailformed or changed
+    // format)
+    DeserializeError(anyhow::Error),
+    Other(anyhow::Error),
+}
+
+impl From<anyhow::Error> for UpdateError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::Other(err)
+    }
+}
+
+impl From<UpdateError> for anyhow::Error {
+    fn from(err: UpdateError) -> Self {
+        match err {
+            UpdateError::SocketClosed(err) => anyhow::anyhow!("socket closed, reason {:?}", err),
+            UpdateError::UnsupportedMessage => anyhow::anyhow!("unsupported message"),
+            UpdateError::DeserializeError(err) => err,
+            UpdateError::Other(err) => err,
+        }
+    }
+}
+
+#[derive(Clone)]
+struct ZeroExCache {
+    /// Fetch orders that can be filled by this address
+    sender: Address,
+    orders: Arc<Mutex<HashMap<OrderId, OrderRecord>>>,
+}
+
+impl ZeroExCache {
+    pub fn new(sender: Address) -> Self {
+        Self {
+            sender,
+            orders: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Calls the 0x API to get all orders and initializes the cache with them.
+    async fn init_cache(&self, api: Arc<dyn ZeroExApi>) -> Result<()> {
+        {
+            // init can be called multiple times, so clear the cache first
+            // clearing is done before fetching 0x orders because fetching can keep failing
+            // for a long time and we don't want to keep stale orders
+            let mut cache = self.orders.lock().await;
+            cache.clear();
```

Comment:
If you don't keep a local variable you don't need the additional `{}` scope.
```suggestion
            let mut cache = self.orders.lock().await.clear();
```